deform/linear_deformation.py

```diff
@@ -0,0 +1,1173 @@
+
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        displacement: `ProductSpaceVector`
+            Linearized deformation parameters for image grid points.
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field. Its space
+            must have the same number of dimensions as ``par_space``.
+        """
+
+        super().__init__(template)
+
+        self.template = template
+        self.displacement = displacement
+
+        Operator.__init__(self, self.displacement.space, self.template.space,
+                          linear=True)
+
+    def derivative(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def _deform_grad(self, grad_f):
+        """Compute the deformation of template's gradient.
+
+        deformed template's gradient: \gradient I(x + v(x))
+
+        Parameters
+        ----------
+        grad_f: `ProductSpaceVector`
+            Gradient of the template, i.e. a vector field on the
+            image domain
+        """
+        temp = [self.linear_deform(gf, self.displacement) for gf in grad_f]
+        return self.displacement.space.element(temp)
+
+    def _call(self, disp_like):
+        """Implementation of ``self(disp_like)``.
+
+        Parameters
+        ----------
+        disp_like: `ProductSpaceVector`
+            Like the linearized deformation parameters for image grid points.
+        """
+        grad = odl.Gradient(self._range)
+        template_grad = grad(self.template)
+        def_grad = self._deform_grad(template_grad)
+
+        result = def_grad[0] * disp_like[0].reshape(self.image_grid.shape,
+                                                    order=self.range.order)
+        for gf, d in zip(def_grad[1:], disp_like[1:]):
+            result += gf * d.reshape(self.image_grid.shape,
+                                     order=self.range.order)
+        return result
+
+    @property
+    def adjoint(self):
+        """Adjoint of the template deformation derivative."""
+        adj_op = LinearizedDeformationDerivativeAdjoint(
+            self.template, self.displacement)
+        return adj_op
+
+
+class LinearizedDeformationDerivativeAdjoint(LinearizedDeformationDerivative):
+
+    """Adjoint of the template deformation operator derivative.
+
+    TODO: write a bit more
+    """
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field. Its space
+            must have the same number of dimensions as ``par_space``.
+        displacement: `ProductSpaceVector`
+            Linearized deformation parameters for image grid points.
+        """
+        super().__init__(template, displacement)
+
+        # Switch domain and range
+        self._domain, self._range = self._range, self._domain
```

Comment:
Additional comment on the above:

It seems we do this exact line in 3 places in the code atm, but i.m.o. the proper way of doing designs like this (if there are things in common) is to create a common base class that both can inherit from.

Currently we have:

`FourierTransformInverse` -> `FourierTransform`

The "proper" design would i.m.o. be to factor out the common code into a helper class

`FourierTransformInverse` -> `FourierTransformBase` <- `FourierTransform`

PEP8 has some info [on this](https://www.python.org/dev/peps/pep-0008/#designing-for-inheritance)

> Another category of attributes are those that are part of the "subclass API" (often called "protected" in other languages). Some classes are designed to be inherited from, either to extend or modify aspects of the class's behavior. When designing such a class, take care to make explicit decisions about which attributes are public, which are part of the subclass API, and which are truly only to be used by your base class.

Perhaps we need to open an issue and have some discussion on this, but my understanding so far has been that unless explicitly stated (such as with the `_call`/`_lincomb` etc methods), private variables are *not* part of the subclass API.